---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MIC

<!-- badges: start -->
[![R-CMD-check](https://github.com/agerada/MIC/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/agerada/MIC/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

## Introduction

`MIC` is an R package for the analysis of minimum inhibitory concentration (MIC)
data. The package was designed to be compatible with the [`AMR`](https://msberends.github.io/AMR/), in
particular most of the functions in `MIC` are designed to accept and return
`AMR` objects, such as `mic` and `sir`. The primary functions in `MIC` are
designed towards validation studies of minimum inhibitory concentrations, however
it also can (optionally) be used to support the construction of machine learning
models that predict MIC values from genomic data.

## Features

- Validation metrics (such as essential agreement) for MIC experiments or
predictions allow comparison against a gold standard, in line with
ISO 20776-2:2021.
- Plots and tables can be generated from validation experiments.
- Quality control analysis of MIC experiments.
- Functions to deal with censoring in MIC data.
- Helper functions to download whole genome sequencing data and susceptibility
metadata from the [PATRIC](https://www.bv-brc.org/docs/system_documentation/data.html)
database at BV-BRC.
- Conversion of whole genome sequence data (assembled .fna files) to k-mer based
features for machine learning models. 
- Fast k-mer counting using C++ and `Rcpp`.
- K-mer features stored in `XGBoost`-compatible `libsvm` format.

## Installation

### CRAN

```r
install.packages("MIC")
```

### GitHub

```r
# install.packages("remotes")
remotes::install_github("agerada/MIC")
```

## Example

Load the `MIC` package -- it is highly recommended that `AMR` is also loaded.
Where possible, `MIC` functions maintain compatibility with `AMR` objects, in
particular the `mic` and `sir` classes.

```{r val-example}
library(MIC)
library(AMR)
```

To compare two `mic` vectors (e.g., one from a gold standard and one from a
prediction or investigational assay), the `compare_mic` function can be used.
An example dataset of MIC values is provided with the package, which will be
used here.

```{r}
data("example_mics")
head(example_mics)
```

The dataset contains MIC values (in `mic` format) for a "test" assay, and a
"gold standard" (`gs`) assay. We will use `compare_mic` to compare the MICs and
validate the "test" assay:

```{r}
val <- compare_mic(gold_standard = example_mics$gs, test = example_mics$test)
val
```

Calling `summary` provides the essential agreement (EA) rates and assay bias:

```{r}
summary(val)
```

If organisms and antimicrobials are provided, `compare_mic` will also calculate
and return the categorical agreement (CA) rates, in the form of minor, major,
and very major errors:

```{r message=FALSE}
val <- compare_mic(gold_standard = example_mics$gs, test = example_mics$test,
                   mo = example_mics$mo, ab = example_mics$ab)
val
```

This time, calling `summary` will provide a breakdown of the categorical
agreement rates in addition to the EA rates:

```{r}
summary(val)
```

Using `as.data.frame` allows us to continue working with the summarised results:

```{r}
head(as.data.frame(val))
```

The results of an `mic_validation` can be plotted in a confusion matrix 
(failed essential agreements are in red):

```{r fig-val-ea, dpi=300}
plot(val)
```

The plot can also be faceted by antimicrobial:

```{r dpi=300, fig.height=14}
plot(val, facet_wrap_ncol = 1)
```

The `table` function can be used to generate a table of the results:

```{r message=FALSE}
# generate table for MEM
mem_dat <- subset(example_mics, ab == "MEM")
mem_val <- compare_mic(gold_standard = mem_dat$gs, test = mem_dat$test)
table(mem_val)
```
